File: src/xmss.c
#include "xmss.h"

/*
 * Compute the root of a Merkle subtree of WOTS+ public keys.
 *
 * Returns length of written output.
 */
int xmss_node(uint8_t *out, const uint8_t sk_seed[ENN], const uint64_t i,
              const uint64_t z, const uint8_t pk_seed[ENN],
              uint32_t adrs[ADRS_LEN]) {
  if (z > HP || i >= (1 << (HP - z)))
    return 0;

  if (z == 0) {
    adrs[ADRS_TYPE_IDX] = WOTS_HASH;
    adrs[ADRS_KEYPAIR_IDX] = htobe32((uint32_t)i);
    bzero(&adrs[ADRS_CHAIN_ADDRESS_IDX], 2 * sizeof(*adrs));

    wotsp_pkgen(out, sk_seed, pk_seed, adrs);
  } else {
    uint8_t buffer[2 * ENN];
    uint8_t *left = buffer;
    uint8_t *right = buffer + ENN;

#ifndef NDEBUG
    int r = xmss_node(left, sk_seed, 2 * i, z - 1, pk_seed, adrs);
    assert(r == ENN);
    r = xmss_node(right, sk_seed, 2 * i + 1, z - 1, pk_seed, adrs);
    assert(r == ENN);
#else
    xmss_node(left, sk_seed, 2 * i, z - 1, pk_seed, adrs);
    xmss_node(right, sk_seed, 2 * i + 1, z - 1, pk_seed, adrs);
#endif

    adrs[ADRS_TYPE_IDX] = TREE;
    adrs[ADRS_KEYPAIR_IDX] = 0; // padding zero for TREE type
    adrs[ADRS_TREE_HEIGHT_IDX] = htobe32((uint32_t)z);
    adrs[ADRS_TREE_INDEX_IDX] = htobe32((uint32_t)i);

    hash_h(out, pk_seed, adrs, buffer);
  }

  return ENN;
}

/*
 * Generate an XMSS signature
 */
int xmss_sign(uint8_t *out, const uint8_t m[ENN], const uint8_t sk_seed[ENN],
              const uint64_t idx, const uint8_t pk_seed[ENN],
              uint32_t adrs[ADRS_LEN]) {
  assert(idx <= (1 << HP));
  size_t j;

  uint8_t *auth = out + (WOTSP_LEN * ENN);
  uint64_t k;
  for (j = 0; j < HP; j++) {
    k = (idx >> j) ^ 0x01;
    xmss_node(auth, sk_seed, k, j, pk_seed, adrs);
    auth += ENN;
  }

  adrs[ADRS_TYPE_IDX] = WOTS_HASH;
  adrs[ADRS_KEYPAIR_IDX] = htobe32((uint32_t)idx);
  bzero(&adrs[ADRS_CHAIN_ADDRESS_IDX], 2 * sizeof(*adrs));

  wotsp_sign(out, m, sk_seed, pk_seed, adrs);

  return 0;
}

/*
 * Compute an XMSS public key from an XMSS signature
 */
int xmss_pk_from_sig(uint8_t *out, const uint64_t idx, const uint8_t *sig,
                     const uint8_t m[ENN], const uint8_t pk_seed[ENN],
                     uint32_t adrs[ADRS_LEN]) {
  adrs[ADRS_TYPE_IDX] = WOTS_HASH;
  adrs[ADRS_KEYPAIR_IDX] = htobe32((uint32_t)idx);
  bzero(&adrs[ADRS_CHAIN_ADDRESS_IDX], 2 * sizeof(*adrs));

  const uint8_t *auth = sig + (WOTSP_LEN * ENN);
  uint8_t node_data[3 * ENN];

  wotsp_pk_from_sig(node_data + ENN, sig, m, ENN, pk_seed, adrs);

  adrs[ADRS_TYPE_IDX] = TREE;
  adrs[ADRS_KEYPAIR_IDX] = 0; // padding zero for TREE type
  adrs[ADRS_TREE_INDEX_IDX] = htobe32((uint32_t)idx);

  size_t k;
  for (k = 0; k < HP; k++) {
    adrs[ADRS_TREE_HEIGHT_IDX] = htobe32((uint32_t)k + 1);
    uint32_t ti = betoh32(adrs[ADRS_TREE_INDEX_IDX]);
    if ((idx >> k) % 2 == 0) {
      adrs[ADRS_TREE_INDEX_IDX] = htobe32(ti >> 1);
      memcpy(node_data + 2 * ENN, auth + k * ENN, ENN);
      hash_h(node_data + ENN, pk_seed, adrs, node_data + ENN);
    } else {
      adrs[ADRS_TREE_INDEX_IDX] = htobe32((ti - 1) >> 1);
      memcpy(node_data, auth + k * ENN, ENN);
      hash_h(node_data + ENN, pk_seed, adrs, node_data);
    }
  }

  memcpy(out, node_data + ENN, ENN);

  return 0;
}

File: python/xmss.py
# From Section 6

from address import Address, AddressType
from slh_dsa import SLHDSA
from utils import toByte
from wotsp import wotsp_pkgen, wotsp_sign, wotsp_pk_from_sig

def xmss_node(sk_seed: bytes, i: int, z: int, pk_seed: bytes, adrs: Address, ctx: SLHDSA) -> bytes:
    """Compute the root of a Merkle subtree of WOTS+ public keys"""
    assert i >= 0
    assert z >= 0

    if z > ctx.hp or i >= 2**(ctx.hp - z):
        return b"" # NULL

    if z == 0:
        adrs.set_type_and_clear(AddressType.WOTS_HASH)
        adrs.set_key_pair_address(toByte(i, 4))
        node = wotsp_pkgen(sk_seed, pk_seed, adrs, ctx)
    else:
        lnode = xmss_node(sk_seed, 2*i    , z-1, pk_seed, adrs, ctx)
        rnode = xmss_node(sk_seed, 2*i + 1, z-1, pk_seed, adrs, ctx)
        adrs.set_type_and_clear(AddressType.TREE)
        adrs.set_tree_height(z)
        adrs.set_tree_index(i)
        node = ctx.hf(pk_seed, bytes(adrs.data), lnode + rnode)
    return node

def xmss_sign(M: bytes, sk_seed: bytes, idx: int, pk_seed: bytes, adrs: Address, ctx: SLHDSA) -> bytes:
    """Generate an XMSS signature"""
    assert len(M) == ctx.n
    assert idx >= 0
    assert idx < (2**ctx.hp)

    auth = bytes()
    for j in range(ctx.hp):
        k = (idx//(2**j)) ^ 0b01
        auth += xmss_node(sk_seed, k, j, pk_seed, adrs, ctx)

    adrs.set_type_and_clear(AddressType.WOTS_HASH)
    adrs.set_key_pair_address(toByte(idx, 4))
    sig = wotsp_sign(M, sk_seed, pk_seed, adrs, ctx)
    sig_xmss = sig + auth
    return sig_xmss

def xmss_pk_from_sig(idx: int, sig_xmss: bytes, M: bytes, pk_seed: bytes, adrs: Address, ctx: SLHDSA) -> bytes:
    """Compute an XMSS public key from an XMSS signature"""
    assert idx >= 0
    assert len(M) == ctx.n
    assert len(sig_xmss) == (ctx.wotsp_len + ctx.hp) * ctx.n

    adrs.set_type_and_clear(AddressType.WOTS_HASH)
    adrs.set_key_pair_address(toByte(idx, 4))
    sig = sig_xmss[:ctx.wotsp_len * ctx.n]  # sig <- sig_xmss.getWOTSSig()
    auth = sig_xmss[ctx.wotsp_len * ctx.n:] # auth <- sig_xmss.getXMSSAUTH()
    node = [wotsp_pk_from_sig(sig, M, pk_seed, adrs, ctx), None]

    adrs.set_type_and_clear(AddressType.TREE)
    adrs.set_tree_index(idx)
    for k in range(ctx.hp):
        adrs.set_tree_height(k + 1)
        if (idx//(2**k)) % 2 == 0:
            adrs.set_tree_index(adrs.get_tree_index()//2)
            node[1] = ctx.hf(pk_seed, bytes(adrs.data), node[0] + auth[k*ctx.n : (k+1)*ctx.n])
        else:
            adrs.set_tree_index((adrs.get_tree_index()-1)//2)
            node[1] = ctx.hf(pk_seed, bytes(adrs.data), auth[k*ctx.n : (k+1)*ctx.n] + node[0])
        node[0] = node[1]
    return node[0]
